C error when using match with option types

### Describe the bug

C error when using match with option types

when there's only part a:
- compiles and runs fine when `none` branch is deleted
- otherwise fails to compile

when there's only part b:
- compiles fine but panics at runtime when `none` branch is deleted
- otherwise fails to compile

part c always fails to compile

### Expected Behavior

compiles fine

### Current Behavior

```
==================
C:/Users/33301/AppData/Local/Temp/v_0/option_match.12583464965555882821.tmp.c:7638: warning: cast between pointer and integer of different size
C:/Users/33301/AppData/Local/Temp/v_0/option_match.12583464965555882821.tmp.c:12659: error: invalid operand types for binary operation
C:/Users/33301/AppData/Local/Temp/v_0/option_match.12583464965555882821.tmp.c:12670: error: cannot convert 'extern struct IError' to 'struct string'
...
==================
(Use `v -cg` to print the entire error message)

builder error:
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang
```
when there's only part b without `none` branch:
```
V panic: string.eq(): nil string
v hash: f1e9a8f
C:/Users/33301/AppData/Local/Temp/v_0/option_match.11156117674075994022.tmp.c:6942: at _v_panic: Backtrace
C:/Users/33301/AppData/Local/Temp/v_0/option_match.11156117674075994022.tmp.c:10165: by string__eq
C:/Users/33301/AppData/Local/Temp/v_0/option_match.11156117674075994022.tmp.c:12641: by main__main
C:/Users/33301/AppData/Local/Temp/v_0/option_match.11156117674075994022.tmp.c:13016: by wmain
0046fb90 : by ???
0046fcf3 : by ???
7ffeb46426bd : by ???
```

### Reproduction Steps

```v
module main

fn main() {
	// part a
	a := ?int(12)
	match a {
		12 { println(a) }
		none { println('none') }
		else { println('else') }
	}
	// part b
	b := ?string(none)
	match b {
		'aaa' { println(b) }
		none { println('none') }
		else { println('else') }
	}
	// part c
	match option_string() {
		'aaa' { println('c') }
		else { println('else') }
	}
}

fn option_string() ?string {
	return none
}
```

### Possible Solution

_No response_

### Additional Information/Context

_No response_

### V version

V 0.3.3 f1e9a8f

### Environment details (OS name and version, etc.)

```
V full version: V 0.3.3 0ba0fb2.f1e9a8f
OS: windows, Microsoft Windows 11 רҵ�� v22621 64 λ
Processor: 12 cpus, 64bit, little endian, 

getwd: C:\Users\33301\Desktop
vexe: E:\Programs\v\v.exe
vexe mtime: 2023-03-19 13:04:16

vroot: OK, value: E:\Programs\v
VMODULES: OK, value: C:\Users\33301\.vmodules
VTMP: OK, value: C:\Users\33301\AppData\Local\Temp\v_0

Git version: git version 2.38.1.windows.1
Git vroot status: weekly.2023.11-37-gf1e9a8ff
.git/config present: true

CC version: Error: exec failed (CreateProcess) with code 2: 系统找不到指定的文件。
 cmd: cc --version
thirdparty/tcc status: thirdparty-windows-amd64 1e6e7c6f
```

--- a/vlib/v/checker/match.v
+++ b/vlib/v/checker/match.v
@@ -29,6 +29,7 @@ fn (mut c Checker) match_expr(mut node ast.MatchExpr) ast.Type {
 	c.ensure_type_exists(node.cond_type, node.pos) or { return ast.void_type }
 	c.check_expr_opt_call(node.cond, cond_type)
 	cond_type_sym := c.table.sym(cond_type)
+	cond_is_option := cond_type.has_flag(.option)
 	node.is_sum_type = cond_type_sym.kind in [.interface_, .sum_type]
 	c.match_exprs(mut node, cond_type_sym)
 	c.expected_type = cond_type
@@ -58,6 +59,10 @@ fn (mut c Checker) match_expr(mut node ast.MatchExpr) ast.Type {
 					c.expected_type = node.expected_type
 				}
 				expr_type := c.expr(stmt.expr)
+				if !branch.is_else && cond_is_option && !expr_type.has_flag(.option) {
+					c.error('`match` expression with Option type only checks against `none`, to match its value you must unwrap it first `var?`',
+						stmt.pos)
+				}
 				stmt.typ = expr_type
 				if first_iteration {
 					if node.is_expr && (node.expected_type.has_flag(.option)

--- a/vlib/v/checker/match.v
+++ b/vlib/v/checker/match.v
@@ -60,9 +60,8 @@ fn (mut c Checker) match_expr(mut node ast.MatchExpr) ast.Type {
 				}
 				expr_type := c.expr(stmt.expr)
 				if !branch.is_else && cond_is_option && branch.exprs[0] !is ast.None {
-					println(c.table.type_to_str(expr_type))
 					c.error('`match` expression with Option type only checks against `none`, to match its value you must unwrap it first `var?`',
-						stmt.pos)
+						branch.pos)
 				}
 				stmt.typ = expr_type
 				if first_iteration {

--- a/vlib/v/checker/match.v
+++ b/vlib/v/checker/match.v
@@ -59,7 +59,8 @@ fn (mut c Checker) match_expr(mut node ast.MatchExpr) ast.Type {
 					c.expected_type = node.expected_type
 				}
 				expr_type := c.expr(stmt.expr)
-				if !branch.is_else && cond_is_option && !expr_type.has_flag(.option) {
+				if !branch.is_else && cond_is_option && branch.exprs[0] !is ast.None {
+					println(c.table.type_to_str(expr_type))
 					c.error('`match` expression with Option type only checks against `none`, to match its value you must unwrap it first `var?`',
 						stmt.pos)
 				}

--- a/vlib/v/checker/tests/match_cond_with_parenthesis_err.out
+++ b/vlib/v/checker/tests/match_cond_with_parenthesis_err.out
@@ -1,7 +1,21 @@
 vlib/v/checker/tests/match_cond_with_parenthesis_err.vv:14:15: error: unnecessary `()` in `match` condition, use `match expr {` instead of `match (expr) {`.
-   12 |
+   12 | 
    13 | fn bar() bool {
    14 |     return match (foo()) {
       |                  ~~~~~~~
    15 |         .a { true }
    16 |         .b, .c { false }
+vlib/v/checker/tests/match_cond_with_parenthesis_err.vv:15:3: error: `match` expression with Option type only checks against `none`, to match its value you must unwrap it first `var?`
+   13 | fn bar() bool {
+   14 |     return match (foo()) {
+   15 |         .a { true }
+      |         ~~
+   16 |         .b, .c { false }
+   17 |     }
+vlib/v/checker/tests/match_cond_with_parenthesis_err.vv:16:3: error: `match` expression with Option type only checks against `none`, to match its value you must unwrap it first `var?`
+   14 |     return match (foo()) {
+   15 |         .a { true }
+   16 |         .b, .c { false }
+      |         ~~~~~~
+   17 |     }
+   18 | }

--- a/vlib/v/checker/tests/option_with_match_err.out
+++ b/vlib/v/checker/tests/option_with_match_err.out
@@ -1,0 +1,7 @@
+vlib/v/checker/tests/option_with_match_err.vv:4:8: error: `match` expression with Option type only checks against `none`, to match its value you must unwrap it first `var?`
+    2 |     mut a := ?int(12)
+    3 |     match a {
+    4 |         12 { println(a) }
+      |              ~~~~~~~~~~
+    5 |         else { println('else') }
+    6 |     }

--- a/vlib/v/checker/tests/option_with_match_err.out
+++ b/vlib/v/checker/tests/option_with_match_err.out
@@ -1,7 +1,7 @@
-vlib/v/checker/tests/option_with_match_err.vv:4:8: error: `match` expression with Option type only checks against `none`, to match its value you must unwrap it first `var?`
+vlib/v/checker/tests/option_with_match_err.vv:4:3: error: `match` expression with Option type only checks against `none`, to match its value you must unwrap it first `var?`
     2 |     mut a := ?int(12)
     3 |     match a {
     4 |         12 { println(a) }
-      |              ~~~~~~~~~~
+      |         ~~
     5 |         else { println('else') }
     6 |     }

--- a/vlib/v/checker/tests/option_with_match_err.vv
+++ b/vlib/v/checker/tests/option_with_match_err.vv
@@ -1,0 +1,7 @@
+fn main() {
+	mut a := ?int(12)
+	match a {
+		12 { println(a) }
+		else { println('else') }
+	}
+}

--- a/vlib/v/gen/c/match.v
+++ b/vlib/v/gen/c/match.v
@@ -65,7 +65,9 @@ fn (mut g Gen) match_expr(node ast.MatchExpr) {
 			g.inside_match_result = true
 		}
 	}
-	if node.cond in [ast.Ident, ast.IntegerLiteral, ast.StringLiteral, ast.FloatLiteral]
+	if (node.cond in [ast.Ident, ast.IntegerLiteral, ast.StringLiteral, ast.FloatLiteral]
+		&& (node.cond !is ast.Ident || (node.cond is ast.Ident
+		&& (node.cond as ast.Ident).or_expr.kind == .absent)))
 		|| (node.cond is ast.SelectorExpr
 		&& (node.cond as ast.SelectorExpr).or_block.kind == .absent
 		&& ((node.cond as ast.SelectorExpr).expr !is ast.CallExpr
@@ -437,6 +439,14 @@ fn (mut g Gen) match_expr_classic(node ast.MatchExpr, is_expr bool, cond_var str
 				if i > 0 {
 					g.write(' || ')
 				}
+				if expr is ast.None {
+					old_left_is_opt := g.left_is_opt
+					g.left_is_opt = true
+					g.expr(node.cond)
+					g.left_is_opt = old_left_is_opt
+					g.write('.state == 2')
+					continue
+				}
 				match type_sym.kind {
 					.array {
 						ptr_typ := g.equality_fn(node.cond_type)
@@ -478,6 +488,12 @@ fn (mut g Gen) match_expr_classic(node ast.MatchExpr, is_expr bool, cond_var str
 							g.write('${cond_var} <= ')
 							g.expr(expr.high)
 							g.write(')')
+						} else if expr is ast.None {
+							old_left_is_opt := g.left_is_opt
+							g.left_is_opt = true
+							g.expr(node.cond)
+							g.left_is_opt = old_left_is_opt
+							g.write('.state == 2')
 						} else {
 							g.write('${cond_var} == (')
 							g.expr(expr)

--- a/vlib/v/tests/option_match_test.v
+++ b/vlib/v/tests/option_match_test.v
@@ -1,0 +1,66 @@
+fn test_simple_match_expr() {
+	mut a := ?int(12)
+	match a? {
+		12 {
+			println(a)
+		}
+		else {
+			println('else')
+			assert false
+		}
+	}
+
+	match a {
+		none {
+			println('none')
+			assert false
+		}
+		else {
+			println('else')
+		}
+	}
+
+	a = none
+
+	match a {
+		none {
+			println('none')
+		}
+		else {
+			println('else')
+			assert false
+		}
+	}
+
+	mut b := ?string('aaa')
+	match b? {
+		'aaa' {
+			println(b)
+		}
+		else {
+			println('else')
+			assert false
+		}
+	}
+
+	match b {
+		none {
+			println('none')
+			assert false
+		}
+		else {
+			println('else')
+		}
+	}
+
+	b = none
+	match b {
+		none {
+			println('none')
+		}
+		else {
+			println('else')
+			assert false
+		}
+	}
+}

